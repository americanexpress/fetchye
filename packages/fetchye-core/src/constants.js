/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

export const ACTION_NAMESPACE = '@fetchye';
export const IS_LOADING = `${ACTION_NAMESPACE}/IS_LOADING`;
export const SET_DATA = `${ACTION_NAMESPACE}/SET_DATA`;
export const DELETE_DATA = `${ACTION_NAMESPACE}/DELETE_DATA`;
export const ERROR = `${ACTION_NAMESPACE}/ERROR`;
export const CLEAR_ERROR = `${ACTION_NAMESPACE}/CLEAR_ERROR`;
