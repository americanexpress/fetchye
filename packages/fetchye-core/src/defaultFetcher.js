/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { trimOptions } from './trimOptions';

// Object.fromEntries is not compatible with Node v10
// provide our own lightweight solution
export const headersToObject = (headers = []) => [...headers]
  .reduce((acc, [key, value]) => {
    acc[key] = value;
    return acc;
  }, {});
export const defaultFetcher = async (fetchClient, key, initialOptions) => {
  let res;
  let payload;
  let error;
  const options = trimOptions(initialOptions);
  try {
    res = await fetchClient(key, options);
    let body = await res.text();
    try {
      body = JSON.parse(body);
    } catch (e) {
      // body will still be the text from the response, so no action needed here
    }
    payload = {
      body,
      ok: res.ok,
      headers: headersToObject(res.headers),
      status: res.status,
    };
  } catch (requestError) {
    // eslint-disable-next-line no-console -- error useful to developer in specific error case
    console.error(requestError);
    error = requestError;
  }
  return {
    payload,
    error,
  };
};

export default defaultFetcher;
