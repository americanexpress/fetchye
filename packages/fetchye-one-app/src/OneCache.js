/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { ImmutableCache } from 'fetchye-immutable-cache';
import invariant from 'invariant';

export const oneCacheSelector = (state) => {
  const rootModuleName = state.getIn(['config', 'rootModuleName']);
  return state.getIn(['modules', rootModuleName, 'fetchye']);
};

export default function OneCache(args = {}) {
  invariant(
    !args.cacheSelector,
    '"cache" is not a permitted argument for OneCache, please use the ImmutableCache provided by fetchye-immutable-cache: https://github.com/americanexpress/fetchye#immutablecache'
  );
  return ImmutableCache({
    cacheSelector: oneCacheSelector,
  });
}
