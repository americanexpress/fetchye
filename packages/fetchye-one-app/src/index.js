export { makeOneServerFetchye } from './makeOneServerFetchye';
export OneCache, { oneCacheSelector } from './OneCache';
export OneFetchyeProvider from './OneFetchyeProvider';
