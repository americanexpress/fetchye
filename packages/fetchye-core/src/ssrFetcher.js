import { defaultFetcher } from './defaultFetcher';

export const ssrFetcher = async (...params) => {
  const { payload } = await defaultFetcher(...params);
  return {
    payload,
    error: null,
  };
};

export default ssrFetcher;
