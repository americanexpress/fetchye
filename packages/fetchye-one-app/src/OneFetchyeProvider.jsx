/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { FetchyeReduxProvider } from 'fetchye-redux-provider';
import invariant from 'invariant';
import OneCache from './OneCache';

const OneFetchyeProvider = ({
  children,
  fetcher,
  fetchClient,
  equalityChecker,
  cache,
}) => {
  invariant(
    !cache,
    '"cache" is not a permitted argument for OneFetchyeProvider, please use the FetchyeReduxProvider provided by fetchye-redux-provider instead: https://github.com/americanexpress/fetchye#fetchyereduxprovider'
  );
  invariant(
    !equalityChecker,
    '"equalityChecker" is not a permitted argument for OneFetchyeProvider, please use the FetchyeReduxProvider provided by fetchye-redux-provider instead: https://github.com/americanexpress/fetchye#fetchyereduxprovider'
  );
  invariant(
    !fetcher,
    '"fetcher" is not a permitted argument for OneFetchyeProvider, please use the FetchyeReduxProvider provided by fetchye-redux-provider instead: https://github.com/americanexpress/fetchye#fetchyereduxprovider'
  );
  return (
    <FetchyeReduxProvider
      cache={OneCache()}
      fetchClient={fetchClient}
    >
      {children}
    </FetchyeReduxProvider>
  );
};

OneFetchyeProvider.propTypes = {
  cache: PropTypes.shape({
    reducer: PropTypes.func.isRequired,
    getCacheByKey: PropTypes.func.isRequired,
    cacheSelector: PropTypes.func.isRequired,
  }).isRequired,
  equalityChecker: PropTypes.func,
  fetchClient: PropTypes.func,
  fetcher: PropTypes.func,
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
};

export default OneFetchyeProvider;
