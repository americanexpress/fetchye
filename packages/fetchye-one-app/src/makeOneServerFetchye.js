import { makeServerFetchye } from 'fetchye';
import OneCache from './OneCache';

export const makeOneServerFetchye = ({
  store,
  fetchClient,
  cache,
}) => makeServerFetchye({
  store,
  fetchClient,
  cache: cache || OneCache(),
});
