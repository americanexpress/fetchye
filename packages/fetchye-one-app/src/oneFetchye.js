import { makeServerFetchye } from 'fetchye';
import OneCache from './OneCache';

const oneFetchye = (...fetchyeArgs) => async (dispatch, getState, { fetchClient }) => {
  const fetchye = makeServerFetchye({
    store: { getState, dispatch },
    fetchClient,
    cache: OneCache(),
  });
  return fetchye(...fetchyeArgs);
};

export default oneFetchye;
