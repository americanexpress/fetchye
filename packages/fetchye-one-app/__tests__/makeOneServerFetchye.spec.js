/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { makeServerFetchye } from 'fetchye';
import { makeOneServerFetchye } from '../src/makeOneServerFetchye';
import OneCache from '../src/OneCache';

jest.mock('../src/OneCache');
jest.mock('fetchye', () => ({
  makeServerFetchye: jest.fn(),
}));

describe('makeOneServerFetchye', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('uses OneCache by default', async () => {
    const store = jest.fn();
    const fetchClient = jest.fn();

    makeOneServerFetchye({
      store,
      fetchClient,
    });

    expect(makeServerFetchye).toHaveBeenCalledWith({
      cache: OneCache(),
      store,
      fetchClient,
    });
  });

  it('throws when cache given', () => {
    const store = jest.fn();
    const fetchClient = jest.fn();
    const cache = jest.fn();

    expect(() => makeOneServerFetchye({
      store,
      fetchClient,
      cache,
    })).toThrowErrorMatchingSnapshot();
  });
});
