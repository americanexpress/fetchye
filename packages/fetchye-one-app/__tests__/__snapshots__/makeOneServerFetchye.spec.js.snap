// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeOneServerFetchye throws when cache given 1`] = `"\\"cache\\" is not a permitted argument for makeOneServerFetchye, please use the makeServerFetchye provided by fetchye instead: https://github.com/americanexpress/fetchye#makeserverfetchye"`;
